File: 37-sudoku-solver/37-sudoku-solver.java
class Solution {
    boolean isValid(int row, int col, char n, char[][] board) {
        for(int i=0; i<board.length; i++) {
            if(board[i][col] == n) return false; 
        }
        for(int i=0; i<board[row].length; i++) {
            if(board[row][i] == n) return false;
        }
        
        int x = (row / 3) * 3;
        int y = (col / 3) * 3;
        
        for(int i=x; i<x+3; i++) {
            for(int j=y; j<y+3; j++) {
                if(i == row || j == col) continue;
                
                if(board[i][j] == n) return false;
            }
        }
        
        return true;
    }
    boolean solve(char[][] board) {
        char[] ch = new char[]{'1','2','3','4','5','6','7','8','9'};
        
        for(int i=0; i<board.length; i++) {
            for(int j=0; j<board[i].length; j++) {
                if(board[i][j] == '.') {
                    for(char c : ch) {
                        if(isValid(i, j, c, board)) {
                            board[i][j] = c;
                            if(solve(board)) return true;
                            board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        
        return true;
    }
    public void solveSudoku(char[][] board) {
        solve(board);
    }
}

File: 37-sudoku-solver/37-sudoku-solver.py
class Solution:
    def valid(self, row: int, col: int, n: str, board: List[List[str]]) -> bool:
        for i in range(0,9):
            if board[i][col] == n:
                return False
        for j in range(0,9):
            if board[row][j] == n:
                return False
        
        x = (row // 3) * 3
        y = (col // 3) * 3
        
        for i in range(x,x+3):
            for j in range(y,y+3):
                if i == row or j == col:
                    continue
                elif(board[i][j] == n):
                    return False
        
        return True
    def solve(self, board: List[List[str]]) -> bool:
        default = ["1","2","3","4","5","6","7","8","9"]
        
        for i in range(0,9):
            for j in range(0,9):
                if board[i][j] == ".":
                    for c in default:
                        if self.valid(i, j, c, board):
                            board[i][j] = c
                            if self.solve(board): 
                                return True
                            board[i][j] = "."
                    return False
                
        return True
    def solveSudoku(self, board: List[List[str]]) -> None:
        self.solve(board)